### PR TITLE
MPRA FG designed sequences metadata

### PR DESCRIPTION
**Pull request by mikelove**
From January 2024, the MPRA FG has worked to create a unified format for specifying the metadata of designed sequences that are used in MPRA experiments. Note that this does not cover STARR-seq, which does not involve designed sequences. It has been covered many times in the MPRA FG, as well as in the [SAP WG](https://drive.google.com/drive/folders/1yULQ8cIvuKCPWWfSZA4ugMnFM-mzOg-K) in March 2024, and in the Characterization WG.

This PR is the first attempt at converting the MPRA FG specification into JSON format.

That is, I've converted this Gdoc

https://docs.google.com/document/d/1ThHgLjMnS2r-vv_4ZHHO9y4K1Qbc2S2WKKUDw__iYXU/edit

into this Google sheet

https://docs.google.com/spreadsheets/d/1nqZL501D1yDNxKGIaEpRszuU6d4BL9mCst3q9bg-7CM/edit?gid=0#gid=0

and into the following JSON file.

### CODE REVIEW

File: src/schemas/table_schemas/designed_sequences.json [hunk 1]
```diff
@@ -0,0 +1,136 @@
...
+                ]
+            }
+        },
+        {
+            "name": "info",
+            "type": "string",
+            "description": "any additional comment or information"
+        }
+    ],
+    "missingValues": [
```
Comment (mingjiecn) [inline src/schemas/table_schemas/designed_sequences.json:131]: In the sample data file, the missing value is NA.

File: src/schemas/table_schemas/designed_sequences.json [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+                    "+",
+                    "-"
+                ]
+            }
+        },
+        {
+            "name": "variant_class",
+            "type": "array",
+            "description": "the class of the variant(s), allowing for multiple variants to be tested in one sequence (haplotypes)",
+            "constraints": {
```
Comment (mingjiecn) [inline src/schemas/table_schemas/designed_sequences.json:94]: this `constraints` should be under `arrayItem`. For example:
`"arrayItem": {
                "type": "string",
                "constraints": {
                    "enum": [
                        "SNP",
                        "indel"
                    ]
                }
            }`

File: src/schemas/table_schemas/designed_sequences.json [hunk 3]
```diff
@@ -0,0 +1,136 @@
...
+                    "SNP",
+                    "indel"
+                ]
+            }
+        },
+        {
+            "name": "variant_pos",
+            "type": "array",
+            "description": "0-based position of the start of the normalized representation of the variant(s). integer within [0, len(sequence) - 1]",
+            "constraints": {
```
Comment (mingjiecn) [inline src/schemas/table_schemas/designed_sequences.json:105]: his constraints should be under arrayItem. For example:
`"arrayItem": { 
                "type": "integer",
                "constraints": { "minimum": 0 }
            }`

File: src/schemas/table_schemas/designed_sequences.json [hunk 4]
```diff
@@ -0,0 +1,136 @@
...
+        {
+            "name": "SPDI",
+            "type": "array",
+            "description": "0-based, validated SPDI representation of the variant(s), e.g. NC_000001.11:25253603:G:A"
+        },
+        {
+            "name": "allele",
+            "type": "array",
+            "description": "the allele of the variant(s) with respect to the referenece chromosome sequence",
+            "constraints": {
```
Comment (mingjiecn) [inline src/schemas/table_schemas/designed_sequences.json:118]:
```
"arrayItem": {
                "type": "string",
                "constraints": {
                    "enum": [
                        "ref",
                        "alt"
                    ]
                }
            }
```

**Review by mingjiecn — CHANGES_REQUESTED** [review]
Several issues:
1. what is the content type of the file for checking? Need to link the content type with the schema so checkfile knows what schema to use: https://github.com/IGVF-DACC/checkfiles/blob/f0c2404c835c154f245e978a4c974e401a51e81c/src/checkfiles/checkfiles.py#L71
2. In the sample file provided, there is a couple issues. For array type, the missing value should be `NA`, not `[NA]`. Also quotation mark is need for array items, for example: `["SNP"] `nor `[SNP]`.
3. The test case should be added to test_checkfiles.py

File: src/schemas/table_schemas/designed_sequences.json (hunk 1 above)
Comment (mikelove) [reply src/schemas/table_schemas/designed_sequences.json:143]: fixed

File: src/schemas/table_schemas/designed_sequences.json (hunk 3 above)
Comment (mikelove) [reply src/schemas/table_schemas/designed_sequences.json]: fixed

File: src/schemas/table_schemas/designed_sequences.json (hunk 4 above)
Comment (mikelove) [reply src/schemas/table_schemas/designed_sequences.json]: fixed

**Comment by mikelove** [comment]
Thanks @mingjiecn -- I'm ready for a re-review.

One small note: while making the test case, it appeared to me that for minimum value requirements, `tabular_file_check()` would only return the first error, not all the errors present in the file.

**Comment by mikelove** [comment]
Considering a small last-minute change: "SNV" instead of "SNP" here.

This is because the single nucleotide changes in the designed sequences do not always correspond to those seen in a population let alone >1% frequency. So this would be more accurate.

**Comment by mingjiecn** [comment]
> Thanks @mingjiecn -- I'm ready for a re-review.
> 
> One small note: while making the test case, it appeared to me that for minimum value requirements, `tabular_file_check()` would only return the first error, not all the errors present in the file.

Checkfiles will check for all errors and it return max of 10 errors by default: https://github.com/IGVF-DACC/checkfiles/blob/63a9af50999794206ba788c68103effb233552c2/src/checkfiles/checkfiles.py#L37

**Review by mingjiecn — APPROVED** [review]

**Comment by mikelove** [comment]
I see. Thanks for the quick review :)

**Pull request merged**